pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java

```diff
@@ -0,0 +1,196 @@
+    // First include the magic header, length field and offsets array.
+    long length = INDEX_ARRAY_START_OFFSET + Integer.BYTES * (byteArrays.length);
+
+    for (byte[] array: byteArrays) {
+      length += array.length;
+    }
+    return length;
+  }
+
+  public static boolean isVarLengthBytesDictBuffer(PinotDataBuffer buffer) {
+    // If the buffer is smaller than header + numElements size, it's not var length dictionary.
+    if (buffer.size() > INDEX_ARRAY_START_OFFSET) {
+      byte[] header = new byte[MAGIC_HEADER_LENGTH];
+      buffer.copyTo(0, header, 0, MAGIC_HEADER_LENGTH);
+
+      if (Arrays.equals(MAGIC_HEADER, header)) {
+        // Also verify that there is a valid numElements value.
+        int numElements = buffer.getInt(MAGIC_HEADER_LENGTH);
+        return numElements > 0;
+      }
+    }
+
+    return false;
+  }
+
+  private void writeHeader() {
+    for (int offset = 0; offset < MAGIC_HEADER_LENGTH; offset++) {
+      _dataBuffer.putByte(offset, MAGIC_HEADER[offset]);
+    }
+  }
+
+  public void init(byte[][] byteArrays) {
+    this._numElements = byteArrays.length;
+
+    writeHeader();
+
+    // Add the number of elements as the first field in the buffer.
+    _dataBuffer.putInt(MAGIC_HEADER_LENGTH, byteArrays.length);
+
+    // Then write the offset of each of the byte array in the data buffer.
+    int nextOffset = INDEX_ARRAY_START_OFFSET;
+    int nextArrayStartOffset = INDEX_ARRAY_START_OFFSET + Integer.BYTES * byteArrays.length;
+    for (byte[] array: byteArrays) {
+      _dataBuffer.putInt(nextOffset, nextArrayStartOffset);
+      nextOffset += Integer.BYTES;
+      nextArrayStartOffset += array.length;
+    }
+
+    // Finally write the byte arrays.
+    nextArrayStartOffset = INDEX_ARRAY_START_OFFSET + Integer.BYTES * byteArrays.length;
+    for (byte[] array: byteArrays) {
+      _dataBuffer.readFrom(nextArrayStartOffset, array);
+      nextArrayStartOffset += array.length;
+    }
+  }
+
+  int getNumElements() {
+    // Lazily initialize the numElements.
+    if (_numElements == 0) {
+      _numElements = _dataBuffer.getInt(MAGIC_HEADER_LENGTH);
+      Preconditions.checkArgument(_numElements > 0);
+    }
+    return this._numElements;
+  }
+
+  @Override
+  public int getInt(int index) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public long getLong(int index) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public float getFloat(int index) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public double getDouble(int index) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public String getUnpaddedString(int index, int numBytesPerValue, byte paddingByte, byte[] buffer) {
+    return StringUtil.decodeUtf8(getBytes(index, numBytesPerValue, buffer));
+  }
+
+  @Override
+  public String getPaddedString(int index, int numBytesPerValue, byte[] buffer) {
```

Comment:
This should throw an exception instead? The callers are expecting a padded string, will any of the caller break if this just returns the raw string?

Reply:
As per my knowledge, the callers of this method are assuming the result is padded only under `paddingByte != 0` but this class will never be used in those cases.

IMO, all this confusion is happening because `ValueReader.java` has the knowledge of padding/non-padding in the interface methods but it shouldn't be. The padding should be implementation detail. I'm planning to cleanup this along with the naming of `FixedByteValueReaderWriter` in another PR so leaving it for now.

Reply:
Hmm, I am not sure about that. IIRC, '\0' is a valid padding char (ie used for padding), and does not mean absence of padding.

Also, it is likely that the exposing the api's (padded vs unpadded) was an optimization to perform operations on padded strings (ie avoid creating new unpadded strings), and strip padding at the end.

Reply:
@mayankshriv I've checked this again and you're correct that the variable length dict should throw exception for this method. Thanks for pointing it.

I've made the change and pushed it.